### PR TITLE
Auth 2

### PR DESCRIPTION
**Pull request by jasiek-net**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,26 +2,8 @@
...
-    }
-  },
-  session: 'secret-session-key',
-  origin: ['http://localhost:4000', 'http://vikinggarage.com'],
-  host: 'localhost',
-  port: 4000,
-}
-```
+#### Add src/config.ts file with configuration
+Check out src/config.ts.example with sample data
```
Comment (michalmikolajczyk) [inline README.md:6]: <3

File: README.md [hunk 2]
```diff
@@ -31,7 +13,12 @@ npm run dev # for development (hot reloading)
 npm start   # for production environment
 ```
 
-### Setup Database with Postgres
+### Docs
```
Comment (michalmikolajczyk) [inline README.md:16]: ❤️

File: src/config.ts.example [hunk 3]
```diff
@@ -0,0 +1,34 @@
+export default {
+
```
Comment (michalmikolajczyk) [inline src/config.ts.example:2]: let's not make new lines before the blocks, just after them (but not after the last block on a level)

File: src/controllers/auth.ts [hunk 4]
```diff
@@ -1,36 +1,107 @@
...
+  UserChange,
+  UserEmail,
   UserLogin,
   UserSignin,
-} from '../models/user'
-import login from '../services/login'
-import signin from '../services/signin'
+  UserToken,
+} from '../models/auth'
+import {
```
Comment (michalmikolajczyk) [inline src/controllers/auth.ts:15]: `import * as auth from '../services/auth'`
?

File: src/tests/auth/change.spec.ts [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+describe('user/change tests', function() {
+
+  before(function() {
+    return cleandb()
+  })
+
+  it('should return error because of token expired', function(done) {
+    chai.request(server)
+    .post('/user/change')
```
Comment (michalmikolajczyk) [inline src/tests/auth/change.spec.ts:20]: indent call chain

File: src/tests/auth/change.spec.ts [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+const should = chai.should()
+chai.use(chaiHttp)
+
+describe('user/change tests', function() {
+
+  before(function() {
+    return cleandb()
+  })
+
+  it('should return error because of token expired', function(done) {
```
Comment (michalmikolajczyk) [inline src/tests/auth/change.spec.ts:18]: `done` is not needed when you `return` a promise within `it`

File: src/tests/auth/change.spec.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+      token: v1(),
+    })
+    .end(function(err, res) {
+      res.should.have.status(400)
+      res.should.be.json
+      res.body.should.have.property('err')
+      res.body.err.should.equal(true)
+      res.body.should.have.property('msg')
+      res.body.msg.should.be.equal('Token expired')
+      done()
```
Comment (michalmikolajczyk) [inline src/tests/auth/change.spec.ts:33]: unnecessary with correct `return`s

File: src/tests/auth/change.spec.ts [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    })
+  })
+
+  it('should changed password successfully', function(done) {
```
Comment (michalmikolajczyk) [inline src/tests/auth/change.spec.ts:37]: cool

File: src/tests/auth/change.spec.ts [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        res.should.be.json
+        res.body.should.have.property('err')
+        res.body.err.should.equal(false)
+        res.body.should.have.property('msg')
+        res.body.msg.should.be.equal('Password changed successfully')
+        done()
+      })
+    })
+  })
+})
```
Comment (michalmikolajczyk) [inline src/tests/auth/change.spec.ts:59]: :heart:

File: src/tests/auth/login.spec.ts [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+      res.should.have.status(200)
+      res.should.be.json
+      res.body.should.have.property('err')
+      res.body.err.should.equal(false)
+      res.body.should.have.property('msg')
+      res.body.msg.should.be.equal('User logged in successfully')
+      done()
+    })
+  })
+})
```
Comment (michalmikolajczyk) [inline src/tests/auth/login.spec.ts:69]: 😍

File: src/tests/auth/reset.spec.ts [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    .send({
+      email: 'wrong.email@gmail.com'
+    })
+    .end(function(err, res) {
+      res.should.have.status(400)
+      res.should.be.json
+      res.body.should.have.property('err')
+      res.body.err.should.equal(true)
+      res.body.should.have.property('msg')
+      res.body.msg.should.be.equal('User with provided email does not exists')
```
Comment (michalmikolajczyk) [inline src/tests/auth/reset.spec.ts:29]: there is a typo in this copy, however the copy should actually be the same for fails and successes.

`Please check the inbox for the e-mail address you provided, for further instructions. If you did not receive an e-mail, please contact us [email link]`

File: src/tests/auth/signin.spec.ts [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+      res.should.have.status(200)
+      res.should.be.json
+      res.body.should.have.property('err')
+      res.body.err.should.equal(false)
+      res.body.should.have.property('msg')
+      res.body.msg.should.be.equal('User created successfully - email sent')
+      done()
+    })
+  })
+})
```
Comment (michalmikolajczyk) [inline src/tests/auth/signin.spec.ts:76]: cool!

File: src/tests/auth/signin.spec.ts [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+      done()
+    })
+  })
+
+  it('should create new user successfully', function(done) {
+    chai.request(server)
+    .post('/user/signin')
+    .send({
+      name: 'Viking',
+      email: 'viking.garage.app+123@gmail.com',
```
Comment (michalmikolajczyk) [inline src/tests/auth/signin.spec.ts:61]: can we rerun the test?

Comment (michalmikolajczyk) [reply src/tests/auth/signin.spec.ts:61]: I guess with cleandb we can

File: src/tests/auth/verify.spec.ts [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+      res.should.be.json
+      res.body.should.have.property('err')
+      res.body.err.should.equal(true)
+      res.body.should.have.property('msg')
+      res.body.msg.should.be.equal('Token expired')
+      done()
+    })
+  })
+
+  it('should verified user successfully', function(done) {
```
Comment (michalmikolajczyk) [inline src/tests/auth/verify.spec.ts:33]: grammar: `verified` --> `verify`

File: src/controllers/auth.ts [hunk 15]
```diff
@@ -1,36 +1,107 @@
...
+  logout,
+  me,
+  resend,
+  reset,
+  signin,
+  verify,
+} from '../services/auth'
 
-@Route('auth')
+@Route('user')
```
Comment (michalmikolajczyk) [inline src/controllers/auth.ts:27]: don't you just hate decorators?

File: src/controllers/index.ts [hunk 16]
```diff
@@ -1,2 +1,5 @@
-import './auth'
-import './helloworld'
+import * as auth from './auth'
+
+export {
+  auth,
```
Comment (michalmikolajczyk) [inline src/controllers/index.ts:4]: trailing comma

File: src/helpers/nodemailer.ts [hunk 17]
```diff
@@ -0,0 +1,40 @@
+import * as nodemailer from 'nodemailer'
```
Comment (michalmikolajczyk) [inline src/helpers/nodemailer.ts:1]: looks cool
please consider how the module actually sends emails

File: src/helpers/passport.ts [hunk 18]
```diff
@@ -1,54 +1,55 @@
...
       })
-      .catch(done)
+      .catch(err => next(null, false, {message: err}))
     }
   ))
+}
 
-  passport.serializeUser(function(user, done) {
-    done(null, user['id'])
+export function authorize(req, res, next) {
```
Comment (michalmikolajczyk) [inline src/helpers/passport.ts:28]: this is not authorization yet, I believe? Just authentication

File: src/helpers/passport.ts [hunk 19]
```diff
@@ -1,54 +1,55 @@
...
+      let payload = {id: user.dataValues.id}
+      let token = jwt.sign(payload, conf.jwt.secret)
+      return resolve({token, user: user.dataValues})
     })
+    .catch(reject)
   })
-
-  app.use(passport.initialize())
-  app.use(passport.session())
 }
```
Comment (michalmikolajczyk) [inline src/helpers/passport.ts:55]: cool

File: src/sequelize/cleandb.ts [hunk 20]
```diff
@@ -0,0 +1,17 @@
+import { User } from './'
+
+export default function cleandb() {
+  return User.sync({ force: true })
+    .then(function() {
```
Comment (michalmikolajczyk) [inline src/sequelize/cleandb.ts:5]: cool indent!!! 👍

File: src/server.ts [hunk 21]
```diff
@@ -1,39 +1,35 @@
...
 app.use(bodyParser.urlencoded({extended: true}))
 app.use(bodyParser.json())
 app.use(methodOverride());
 
 passportConfig(app)
 RegisterRoutes(app)
 
+app.use(function(err, req, res, next) {
+  console.log(`Unexpected error ${err}`)
+  res.status(500).send(`Unexpected error ${err}`)
```
Comment (michalmikolajczyk) [inline src/server.ts:30]: is there a stack trace here?

File: src/services/auth/change.ts [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+export default function change(req, res, next) {
+
+  let {
+    password1,
+    password2,
+    token
+  } = req.body
+
+  User.findOne({where: { token }})
+  .then(user => {
```
Comment (michalmikolajczyk) [inline src/services/auth/change.ts:13]: indent chain pls

File: src/services/auth/change.ts [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+    .then(() => {
+        res.status(200).json({
+        err: false,
+        msg: 'Password changed successfully'
+      })
+    })
+    .catch(err => {
+      res.status(500).json({
+        err: true,
+        msg: `Unexpected error ${err}`
```
Comment (michalmikolajczyk) [inline src/services/auth/change.ts:34]: what info are we sending to the client?

File: src/services/auth/login.ts [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+      user: {
+        name: user.name,
+        email: user.email,
+      }
+    })
+  })
+  .catch(err => {
+    res.status(400).json({
+      err: true,
+      msg: `Invalid email or password`,
```
Comment (michalmikolajczyk) [inline src/services/auth/login.ts:32]: cool

File: src/services/auth/me.ts [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+  if (req.isAuthenticated()) {
+    res.status(200).json({
+      err: false,
+      msg: 'User is logged in :)',
+      user: req.user
+    })
+  } else {
+    res.json({
+      err: false,
+      msg: 'User is not logged in :(',
```
Comment (michalmikolajczyk) [inline src/services/auth/me.ts:13]: poker face

File: src/services/auth/me.ts [hunk 26]
```diff
@@ -0,0 +1,17 @@
+import { User } from '../../sequelize'
+
+export default function me(req, res, next) {
+  if (req.isAuthenticated()) {
+    res.status(200).json({
+      err: false,
+      msg: 'User is logged in :)',
```
Comment (michalmikolajczyk) [inline src/services/auth/me.ts:7]: let's keep it cool

File: src/services/auth/reset.ts [hunk 27]
```diff
@@ -0,0 +1,37 @@
+import { User } from '../../sequelize'
+import { resetEmail } from '../../helpers/nodemailer'
+
+export default function reset(req, res, next) {
+
+  let { email } = req.body
+
+  User.findOne({where: { email }})
+  .then(user => {
```
Comment (michalmikolajczyk) [inline src/services/auth/reset.ts:9]: indent chain pls

**Review by michalmikolajczyk — APPROVED** [review]
cool

File: src/controllers/auth.ts (hunk 15 above)
Comment (jasiek-net) [reply src/controllers/auth.ts:27]: Yep, I hate, but here decorators allows us to generate openAPI specification straight from controllers (via `tsoa`), so I treat this as a lesser evil.

File: src/helpers/nodemailer.ts (hunk 17 above)
Comment (jasiek-net) [reply src/helpers/nodemailer.ts:1]: it uses `SMTP` as described in [docs](https://nodemailer.com/about/)

File: src/tests/auth/change.spec.ts (hunk 6 above)
Comment (jasiek-net) [reply src/tests/auth/change.spec.ts:18]: it's needed by [chai-http](https://github.com/chaijs/chai-http)

**Pull request merged**